

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>pylprotpredictor.predict &mdash; PylProtPredictor 0.1.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="../../genindex.html"/>
        <link rel="search" title="Search" href="../../search.html"/>
    <link rel="top" title="PylProtPredictor 0.1.0 documentation" href="../../index.html"/>
        <link rel="up" title="Module code" href="../index.html"/> 

  
  <script src="../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../index.html" class="icon icon-home"> PylProtPredictor
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">PylProtPredictor documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../pylprotpredictor.html">Identification of potential PYL proteins</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../use_case.html">Some example of usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../code.html">Code: pylprotpredictor module</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/bebatut/PylProtPredictor">Source code on GitHub</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">PylProtPredictor</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../index.html">Module code</a> &raquo;</li>
        
      <li>pylprotpredictor.predict</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for pylprotpredictor.predict</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">import</span> <span class="nn">logging</span>

<span class="kn">from</span> <span class="nn">Bio</span> <span class="k">import</span> <span class="n">SeqIO</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="k">import</span> <span class="n">Path</span>

<span class="k">try</span><span class="p">:</span>
    <span class="kn">from</span> <span class="nn">pylprotpredictor</span> <span class="k">import</span> <span class="n">cds</span>
    <span class="kn">from</span> <span class="nn">pylprotpredictor</span> <span class="k">import</span> <span class="n">export</span>
<span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
    <span class="kn">from</span> <span class="nn">.</span> <span class="k">import</span> <span class="n">cds</span>
    <span class="kn">from</span> <span class="nn">.</span> <span class="k">import</span> <span class="n">export</span>


<div class="viewcode-block" id="extract_seqs"><a class="viewcode-back" href="../../code.html#pylprotpredictor.predict.extract_seqs">[docs]</a><span class="k">def</span> <span class="nf">extract_seqs</span><span class="p">(</span><span class="n">seq_filepath</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Extract the sequences in a fasta file</span>

<span class="sd">    :param seq_filepath: path to a fasta file</span>

<span class="sd">    :return: a dictionary with all sequences indexed by their id, their length and their complement sequence</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">seqs</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">seq_filepath</span><span class="o">.</span><span class="n">as_posix</span><span class="p">(),</span> <span class="s2">&quot;fasta&quot;</span><span class="p">):</span>
        <span class="n">seqs</span><span class="p">[</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="n">record</span></div>
    <span class="k">return</span> <span class="n">seqs</span>


<div class="viewcode-block" id="extract_predicted_cds"><a class="viewcode-back" href="../../code.html#pylprotpredictor.predict.extract_predicted_cds">[docs]</a><span class="k">def</span> <span class="nf">extract_predicted_cds</span><span class="p">(</span>
        <span class="n">pred_cds_path</span><span class="p">,</span> <span class="n">pred_cds_info_path</span><span class="p">,</span> <span class="n">tag_ending_cds_info_path</span><span class="p">,</span>
        <span class="n">genome_filepath</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Extract the list of predicted CDS and identify the CDS ending with TAG STOP codon</span>

<span class="sd">    :param pred_cds_path: path to the output of CDS prediction (Prodigal)</span>
<span class="sd">    :param pred_cds_info_path: path to a CSV file in which the information (start, end, strand, origin) are collected for each predicted CDS</span>
<span class="sd">    :param tag_ending_cds_info_path: path to CSV file to export the information about the TAG ending CDS</span>
<span class="sd">    :param genome_filepath: path to reference genome</span>

<span class="sd">    :return: a dictionary with the predicted CDS represented by CDS object</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">pred_cds</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">pred_cds_info</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">tag_ending_cds_info</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="n">origin_seqs</span> <span class="o">=</span> <span class="n">extract_seqs</span><span class="p">(</span><span class="n">genome_filepath</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">pred_cds_path</span><span class="o">.</span><span class="n">as_posix</span><span class="p">(),</span> <span class="s2">&quot;fasta&quot;</span><span class="p">):</span>
        <span class="n">cds_obj</span> <span class="o">=</span> <span class="n">cds</span><span class="o">.</span><span class="n">CDS</span><span class="p">()</span>
        <span class="n">cds_obj</span><span class="o">.</span><span class="n">init_from_record</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
        <span class="n">cds_id</span> <span class="o">=</span> <span class="n">cds_obj</span><span class="o">.</span><span class="n">get_id</span><span class="p">()</span>
        <span class="n">origin_seq_id</span> <span class="o">=</span> <span class="n">cds_obj</span><span class="o">.</span><span class="n">get_origin_seq_id</span><span class="p">()</span>
        <span class="n">strand</span> <span class="o">=</span> <span class="n">cds_obj</span><span class="o">.</span><span class="n">get_strand</span><span class="p">()</span>
        <span class="n">start</span> <span class="o">=</span> <span class="n">cds_obj</span><span class="o">.</span><span class="n">get_start</span><span class="p">()</span>
        <span class="n">end</span> <span class="o">=</span> <span class="n">cds_obj</span><span class="o">.</span><span class="n">get_end</span><span class="p">()</span>
        <span class="n">stop_codon</span> <span class="o">=</span> <span class="n">cds_obj</span><span class="o">.</span><span class="n">get_stop_codon</span><span class="p">()</span>

        <span class="n">pred_cds</span><span class="p">[</span><span class="n">cds_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">cds_obj</span>

        <span class="n">pred_cds_info</span><span class="p">[</span><span class="n">cds_id</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;start&quot;</span><span class="p">:</span> <span class="n">start</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">:</span> <span class="n">end</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="n">strand</span><span class="p">,</span> <span class="s2">&quot;origin_seq&quot;</span><span class="p">:</span> <span class="n">origin_seq_id</span><span class="p">,</span> <span class="s2">&quot;stop_codon&quot;</span><span class="p">:</span> <span class="n">stop_codon</span><span class="p">}</span>

        <span class="k">if</span> <span class="n">cds_obj</span><span class="o">.</span><span class="n">is_tag_ending_seq</span><span class="p">():</span>
            <span class="n">tag_ending_cds_info</span><span class="p">[</span><span class="n">cds_id</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;start&quot;</span><span class="p">:</span> <span class="n">start</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">:</span> <span class="n">end</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="n">strand</span><span class="p">,</span> <span class="s2">&quot;origin_seq&quot;</span><span class="p">:</span> <span class="n">origin_seq_id</span><span class="p">}</span>
            <span class="n">cds_obj</span><span class="o">.</span><span class="n">set_origin_seq</span><span class="p">(</span><span class="n">origin_seqs</span><span class="p">[</span><span class="n">origin_seq_id</span><span class="p">])</span>
            <span class="n">cds_obj</span><span class="o">.</span><span class="n">set_status</span><span class="p">(</span><span class="s2">&quot;tag-ending&quot;</span><span class="p">)</span>

    <span class="n">export</span><span class="o">.</span><span class="n">export_csv</span><span class="p">(</span><span class="n">pred_cds_info</span><span class="p">,</span> <span class="n">pred_cds_info_path</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;origin_seq&quot;</span><span class="p">,</span> <span class="s2">&quot;stop_codon&quot;</span><span class="p">])</span>
    <span class="n">export</span><span class="o">.</span><span class="n">export_csv</span><span class="p">(</span><span class="n">tag_ending_cds_info</span><span class="p">,</span> <span class="n">tag_ending_cds_info_path</span><span class="p">,</span> <span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;origin_seq&quot;</span><span class="p">])</span>

    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Number of predicted CDS: </span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">pred_cds</span><span class="o">.</span><span class="n">keys</span><span class="p">())))</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Number of TAG-ending predicted CDS: </span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">tag_ending_cds_info</span><span class="o">.</span><span class="n">keys</span><span class="p">())))</span>
</div>
    <span class="k">return</span> <span class="n">pred_cds</span>


<div class="viewcode-block" id="extract_potential_pyl_cds"><a class="viewcode-back" href="../../code.html#pylprotpredictor.predict.extract_potential_pyl_cds">[docs]</a><span class="k">def</span> <span class="nf">extract_potential_pyl_cds</span><span class="p">(</span>
        <span class="n">pred_cds</span><span class="p">,</span> <span class="n">pot_pyl_cds_filepath</span><span class="p">,</span> <span class="n">pot_pyl_cds_info_filepath</span><span class="p">,</span> <span class="n">pred_cds_obj_filepath</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Extract potential PYL CDS from TAG-ending CDS</span>

<span class="sd">    :param pred_cds: a dictionary with the predicted CDS represented as CDS objects</span>
<span class="sd">    :param pot_pyl_cds_filepath: path to fasta file in which the protein sequences of the potential PYL CDS are saved</span>
<span class="sd">    :param pot_pyl_cds_info_filepath: path to a cvs file to get information about potential PYL CDS</span>
<span class="sd">    :param pred_cds_obj_filepath: path to generated JSON file to store the list of predicted CDS objects</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">info</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">sequences</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">json_pred_cds</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">cds_id</span> <span class="ow">in</span> <span class="n">pred_cds</span><span class="p">:</span>
        <span class="n">cds_obj</span> <span class="o">=</span> <span class="n">pred_cds</span><span class="p">[</span><span class="n">cds_id</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">cds_obj</span><span class="o">.</span><span class="n">is_tag_ending</span><span class="p">():</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">CDS: </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">cds_id</span><span class="p">))</span>

            <span class="n">cds_obj</span><span class="o">.</span><span class="n">find_alternative_ends</span><span class="p">()</span>
            <span class="n">cds_obj</span><span class="o">.</span><span class="n">extract_possible_alternative_seq</span><span class="p">()</span>

            <span class="k">if</span> <span class="n">cds_obj</span><span class="o">.</span><span class="n">has_alternative_cds</span><span class="p">():</span>
                <span class="n">info</span><span class="p">[</span><span class="n">cds_id</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
                    <span class="s2">&quot;start&quot;</span><span class="p">:</span> <span class="n">cds_obj</span><span class="o">.</span><span class="n">get_start</span><span class="p">(),</span>
                    <span class="s2">&quot;end&quot;</span><span class="p">:</span> <span class="n">cds_obj</span><span class="o">.</span><span class="n">get_end</span><span class="p">(),</span>
                    <span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="n">cds_obj</span><span class="o">.</span><span class="n">get_strand</span><span class="p">(),</span>
                    <span class="s2">&quot;origin_seq&quot;</span><span class="p">:</span> <span class="n">cds_obj</span><span class="o">.</span><span class="n">get_origin_seq_id</span><span class="p">(),</span>
                    <span class="s2">&quot;alternative_start&quot;</span><span class="p">:</span> <span class="s2">&quot;;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">cds_obj</span><span class="o">.</span><span class="n">get_alternative_start</span><span class="p">()]),</span>
                    <span class="s2">&quot;alternative_end&quot;</span><span class="p">:</span> <span class="s2">&quot;;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">cds_obj</span><span class="o">.</span><span class="n">get_alternative_end</span><span class="p">()])}</span>
                <span class="n">sequences</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cds_obj</span><span class="o">.</span><span class="n">get_translated_seq</span><span class="p">())</span>
                <span class="n">sequences</span> <span class="o">+=</span> <span class="n">cds_obj</span><span class="o">.</span><span class="n">get_translated_alternative_seq</span><span class="p">()</span>
                <span class="n">cds_obj</span><span class="o">.</span><span class="n">set_status</span><span class="p">(</span><span class="s2">&quot;potential pyl&quot;</span><span class="p">)</span>

        <span class="n">json_pred_cds</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">pred_cds</span><span class="p">[</span><span class="n">cds_id</span><span class="p">]</span><span class="o">.</span><span class="n">export_to_dict</span><span class="p">())</span>

    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Number of potential Pyl proteins: </span><span class="si">%s</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">info</span><span class="o">.</span><span class="n">keys</span><span class="p">())))</span>
    <span class="n">export</span><span class="o">.</span><span class="n">export_fasta</span><span class="p">(</span><span class="n">sequences</span><span class="p">,</span> <span class="n">pot_pyl_cds_filepath</span><span class="p">)</span>
    <span class="n">export</span><span class="o">.</span><span class="n">export_csv</span><span class="p">(</span>
        <span class="n">info</span><span class="p">,</span>
        <span class="n">pot_pyl_cds_info_filepath</span><span class="p">,</span>
        <span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">,</span> <span class="s2">&quot;end&quot;</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">,</span> <span class="s2">&quot;origin_seq&quot;</span><span class="p">,</span> <span class="s2">&quot;alternative_start&quot;</span><span class="p">,</span> <span class="s2">&quot;alternative_end&quot;</span><span class="p">])</span></div>
    <span class="n">export</span><span class="o">.</span><span class="n">export_json</span><span class="p">(</span><span class="n">json_pred_cds</span><span class="p">,</span> <span class="n">pred_cds_obj_filepath</span><span class="p">)</span>


<div class="viewcode-block" id="predict_pyl_proteins"><a class="viewcode-back" href="../../code.html#pylprotpredictor.predict.predict_pyl_proteins">[docs]</a><span class="k">def</span> <span class="nf">predict_pyl_proteins</span><span class="p">(</span>
        <span class="n">genome_filepath</span><span class="p">,</span> <span class="n">pred_cds_filepath</span><span class="p">,</span> <span class="n">pot_pyl_seq_filepath</span><span class="p">,</span> <span class="n">log_filepath</span><span class="p">,</span>
        <span class="n">pred_cds_info_filepath</span><span class="p">,</span> <span class="n">tag_ending_cds_info_filepath</span><span class="p">,</span>
        <span class="n">pot_pyl_seq_info_filepath</span><span class="p">,</span> <span class="n">pred_cds_obj_filepath</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Run prediction of potentila PYL CDS:</span>

<span class="sd">    - Extraction of predicted CDS into a dictionary</span>
<span class="sd">    - Identification of TAG-ending proteins</span>
<span class="sd">    - Extraction of potential PYL sequences</span>

<span class="sd">    :param genome_filepath: path to file with genome sequence</span>
<span class="sd">    :param pred_cds_filepath: path to the output of CDS prediction (Prodigal)</span>
<span class="sd">    :param pot_pyl_seq_filepath: path to fasta file with potential PYL CDS sequence</span>
<span class="sd">    :param log_filepath: path to log file</span>
<span class="sd">    :param pred_cds_info_filepath: path to CSV file with predicted CDS info</span>
<span class="sd">    :param tag_ending_cds_info_filepath: path to CSV file with TAG-ending CDS info</span>
<span class="sd">    :param pot_pyl_seq_info_filepath: path to CSV file with potential PYL CDS info</span>
<span class="sd">    :param pred_cds_obj_filepath: path to generated JSON file to store the list of predicted CDS objects</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="n">log_filepath</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">,</span> <span class="n">filemode</span><span class="o">=</span><span class="s2">&quot;w&quot;</span><span class="p">)</span>
    <span class="n">pred_cds</span> <span class="o">=</span> <span class="n">extract_predicted_cds</span><span class="p">(</span>
        <span class="n">pred_cds_filepath</span><span class="p">,</span>
        <span class="n">pred_cds_info_filepath</span><span class="p">,</span>
        <span class="n">tag_ending_cds_info_filepath</span><span class="p">,</span>
        <span class="n">genome_filepath</span><span class="p">)</span>
    <span class="n">extract_potential_pyl_cds</span><span class="p">(</span>
        <span class="n">pred_cds</span><span class="p">,</span>
        <span class="n">pot_pyl_seq_filepath</span><span class="p">,</span>
        <span class="n">pot_pyl_seq_info_filepath</span><span class="p">,</span></div>
        <span class="n">pred_cds_obj_filepath</span><span class="p">)</span>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="s1">&#39;Predict potential PYL CDS&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--genome&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;path to FASTA file with genome (whole or scaffold) on which predict proteins&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--pred_cds&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;path to FASTA file with predicted CDS of the genome&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--pot_pyl_seq&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;path to generated FASTA file with potential PYL CDS sequence&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--log&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;path to generated log file&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--pred_cds_info&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;path to generated CSV file with predicted CDS info&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--pred_cds_obj&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;path to generated JSON file to store the list of predicted CDS object&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--tag_ending_cds_info&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;path to generated CSV file with TAG-ending CDS info&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--pot_pyl_seq_info&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;path to generated CSV file with potential PYL CDS info&#39;</span><span class="p">)</span>

    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>

    <span class="n">predict_pyl_proteins</span><span class="p">(</span>
        <span class="n">genome_filepath</span><span class="o">=</span><span class="n">Path</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">genome</span><span class="p">),</span>
        <span class="n">pred_cds_filepath</span><span class="o">=</span><span class="n">Path</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">pred_cds</span><span class="p">),</span>
        <span class="n">pot_pyl_seq_filepath</span><span class="o">=</span><span class="n">Path</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">pot_pyl_seq</span><span class="p">),</span>
        <span class="n">log_filepath</span><span class="o">=</span><span class="n">Path</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">log</span><span class="p">),</span>
        <span class="n">pred_cds_info_filepath</span><span class="o">=</span><span class="n">Path</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">pred_cds_info</span><span class="p">),</span>
        <span class="n">tag_ending_cds_info_filepath</span><span class="o">=</span><span class="n">Path</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">tag_ending_cds_info</span><span class="p">),</span>
        <span class="n">pot_pyl_seq_info_filepath</span><span class="o">=</span><span class="n">Path</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">pot_pyl_seq_info</span><span class="p">),</span>
        <span class="n">pred_cds_obj_filepath</span><span class="o">=</span><span class="n">Path</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">pred_cds_obj</span><span class="p">))</span>
</pre></div>

           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Bérénice Batut, Kévin Gravouil, Jean-Francois Brugère.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../',
            VERSION:'0.1.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  false,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>